python/ray/_private/gcs_utils.py

```diff
@@ -182,13 +196,227 @@ def internal_kv_keys(self, prefix: bytes) -> List[bytes]:
+            key_id=key_id,
+            error_info_message=error_info)
+        req = gcs_service_pb2.GcsPublishRequest(pub_messages=[msg])
+        await self._stub.GcsPublish(req)
+
+
+class GcsAioSubscriber:
+    """Subscriber to GCS. Uses async io."""
+
+    def __init__(self, address: str = None, channel: grpc.aio.Channel = None):
+        if address:
+            assert channel is None, \
+                "address and channel cannot both be specified"
+            channel = create_gcs_aio_channel(address)
+        else:
+            assert channel is not None, \
+                "One of address and channel must be specified"
+        self._stub = gcs_service_pb2_grpc.InternalPubSubGcsServiceStub(channel)
+        self._subscriber_id = bytes(
+            bytearray(random.getrandbits(8) for _ in range(28)))
+        # Whether error info has been subscribed.
+        self._subscribed_error = False
+        # Buffer for holding error info.
+        self._errors = deque()
+
+    async def subscribe_error(self) -> None:
+        """Starts a subscription for error info"""
+        cmd = pubsub_pb2.Command(
+            channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+            subscribe_message={})
+        req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
+            subscriber_id=self._subscriber_id, commands=[cmd])
+        await self._stub.GcsSubscriberCommandBatch(req)
+        self._subscribed_error = True
+
+    async def poll_error(self, timeout=None) -> Tuple[bytes, ErrorTableData]:
+        """Polls for new error messages."""
+        if not self._subscribed_error:
+            self.subscribe_error()
+
+        if len(self._errors) == 0:
+            req = gcs_service_pb2.GcsSubscriberPollRequest(
+                subscriber_id=self._subscriber_id)
+            reply = await self._stub.GcsSubscriberPoll(req, timeout=timeout)
+            for msg in reply.pub_messages:
+                self._errors.append((msg.key_id, msg.error_info_message))
+
+        if len(self._errors) == 0:
+            return None, None
+        return self._errors.popleft()
+
+        req = gcs_service_pb2.GcsSubscriberPollRequest(
+            subscriber_id=self._subscriber_id)
+        reply = await self._stub.GcsSubscriberPoll(req)
+        error_info = []
+        for msg in reply.pub_messages:
+            error_info.append((msg.key_id, msg.error_info_message))
+        return error_info
+
+    async def close(self) -> None:
+        """Closes the subscriber and its active subscriptions."""
+        req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
+            subscriber_id=self._subscriber_id, commands=[])
+        if self._subscribed_error:
+            cmd = pubsub_pb2.Command(
+                channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+                unsubscribe_message={})
+            req.commands.append(cmd)
+        try:
+            await self._stub.GcsSubscriberCommandBatch(req, timeout=30)
```

Comment:
Subscriber commands are usually made on process start and end. If it blocks or fails, Ray cannot start properly. IMO using a deadline here compared to not using a deadline can make issues more apparent, e.g. exception for connection failure instead of process hanging. A deadline too small may introduce flakiness if GCS does not respond in time. The default has been no deadline in Ray, e.g. C++ GCS client calls. I think using a longer deadline is safer than a short deadline for now?